### DISCUSSION TITLE
Inputting rows from an external file, into longtblr table. Getting Misplaced alignment tab character &.

### DISCUSSION THREAD

**Discussion by russurquhart1** [lvjr/tabularray#496] [Q&A]
I'm trying to reuse rows, from another file, and have those rows added to an existing table.  My main file is like this:

```tex
\documentclass[landscape]{article}
\usepackage{tabularray}
\usepackage{xcolor}
\usepackage[hyperindex,hidelinks]{hyperref}
\begin{document}
\begin{longtblr}[
   caption = {Severity Categories}
]{
  %repeat the first row (the header) for each time the table breaks a page
  rowhead = 1,
  colspec = { |Q[c,m,wd=1.0in]|Q[l,m,wd=6.0in]| },
  rows={ rowsep=9pt },
  % set "row{1}" after "colspec" to overwrite "halign"
  row{1}={ gray!10, halign=c, font=\bfseries },
}
  \hline
  Severity & Description \\
  \hline
  a & b \\ \hline
\input{in-file.tex}
\end{longtblr}
\subsection{Test Network} \label{SIL-3079}
\end{document}

```
In the in-file.tex I have the following:

`\hyperref[SIL-3079]{SIL-3079} & \ref{SIL-3079}  \\ \hline
`

I get the following error:

```
LaTeX Warning: Hyper reference `SIL-3079' on page 1 undefined on input line 1.

! Misplaced alignment tab character &.
l.1 \hyperref[SIL-3079]{SIL-3079} &
                                    \ref{SIL-3079}  \\ \hline
? 

```
If i explicitly put the row in the main file, it all works, but i was hoping I'd be able to just have an input file of rows, and use \input to put it in the table.
Does longtblr not allow this?

Any help is greatly appreciated!

Thanks,

Russ

**Comment by muzimuzhi** [comment]
Check out package manual (v2024A, 2024-02-16), sec. 5.5.1 "Outer key `evaluate` in action".

**Comment by russurquhart1** [comment]
I'm still getting the error, I updated the LaTeX source accordingly:
```

\documentclass[landscape]{article}
\usepackage{tabularray}
\usepackage{xcolor}
\usepackage[hyperindex,hidelinks]{hyperref}
\UseTblrLibrary{functional}

\begin{document}
\begin{longtblr}[
   evaluate=fileInput
   caption = {Severity Categories}
]{
  %repeat the first row (the header) for each time the table breaks a page
  rowhead = 1,
  colspec = { |Q[c,m,wd=1.0in]|Q[l,m,wd=6.0in]| },
  rows={ rowsep=9pt },
  % set "row{1}" after "colspec" to overwrite "halign"
  row{1}={ gray!10, halign=c, font=\bfseries },
}
  \hline
  Severity & Description \\
  \hline
  a & b \\ \hline
\fileInput{in-file.tex}
\end{longtblr}
\subsection{Test Network} \label{SIL-3079}
\end{document}
```
Is there something else I need to do?

Thanks,

**Reply by muzimuzhi** [reply]
> ```
> \begin{longtblr}[
>    evaluate=fileInput
>    caption = {Severity Categories}
> ]{
> ```

Replace `evaluate=fileInput` with `evaluate=\fileInput,` then your example compiles. Note the backslash ` \ ` in `\fileInput` and trailing comma for separating key-value pairs.

**Reply by russurquhart1** [reply]
I still must be missing something, as I'm still getting the error. Here is what I currently have:

```
\documentclass[landscape]{article}
\usepackage{tabularray}
\usepackage{xcolor}
\usepackage[hyperindex,hidelinks]{hyperref}
\UseTblrLibrary{functional}

\begin{document}
\begin{longtblr}[
   evaluate=\fileInput,
   caption = {Severity Categories}
]{
  %repeat the first row (the header) for each time the table breaks a page
  rowhead = 1,
  colspec = { |Q[c,m,wd=1.0in]|Q[l,m,wd=6.0in]| },
  rows={ rowsep=9pt },
  % set "row{1}" after "colspec" to overwrite "halign"
  row{1}={ gray!10, halign=c, font=\bfseries },
}
  \hline
  Severity & Description \\
  \hline
  a & b \\ \hline
\fileInput{in-file.tex}
\end{longtblr}
\subsection{Test Network} \label{SIL-3079}
\end{document}
```

**Reply by muzimuzhi** [reply]
What's the full message of the first error?

Your version of `tabularray` may also matter, which you can get near the end of log file after adding `\listfiles` to the preamble.

Your last example plus the `filecontents*` environment compiles both on my side and in Overleaf's TeX Live 2023 with 0 errors, 2 warnings:

```
LaTeX Warning: Writing or overwriting file `in-file.tex'.

Package tabularray Warning: Table width is too small, need 186.08997pt more!
```

```tex
\begin{filecontents*}[overwrite]{in-file.tex}
\hyperref[SIL-3079]{SIL-3079} & \ref{SIL-3079}  \\ \hline
\end{filecontents*}

\documentclass[landscape]{article}
\usepackage{tabularray}
\usepackage{xcolor}
\usepackage[hyperindex,hidelinks]{hyperref}
\UseTblrLibrary{functional}

\begin{document}
\begin{longtblr}[
   evaluate=\fileInput,
   caption = {Severity Categories}
]{
  %repeat the first row (the header) for each time the table breaks a page
  rowhead = 1,
  colspec = { |Q[c,m,wd=1.0in]|Q[l,m,wd=6.0in]| },
  rows={ rowsep=9pt },
  % set "row{1}" after "colspec" to overwrite "halign"
  row{1}={ gray!10, halign=c, font=\bfseries },
}
  \hline
  Severity & Description \\
  \hline
  a & b \\ \hline
  \fileInput{in-file.tex}
\end{longtblr}
\subsection{Test Network} \label{SIL-3079}
\end{document}
```

**Reply by russurquhart1** [reply]
The full message is:

```
(C:/Users/russell.urquhart/Desktop\in-file.tex)

LaTeX Warning: Writing or overwriting file `in-file.tex'.


LaTeX Warning: Hyper reference `SIL-3079' on page 1 undefined on input line 26.


! Misplaced alignment tab character &.
\l__tblr_c_tl ... \hyperref [SIL-3079]{SIL-3079} &
                                                   \ref {SIL-3079} \\ \hline...
l.26 \end
         {longtblr}
? 
```

I added `\listfiles` to the preamble but couldn't find ref to the version of `tabularray` in the Console output or Error, warnings, Badboxes window.

```
\documentclass[landscape]{article}
\usepackage{tabularray}
\usepackage{xcolor}
\usepackage[hyperindex,hidelinks]{hyperref}
\UseTblrLibrary{functional}
\listfiles

\begin{document}
\begin{longtblr}[
   evaluate=\fileInput,
   caption = {Severity Categories}
]{
  %repeat the first row (the header) for each time the table breaks a page
  rowhead = 1,
  colspec = { |Q[c,m,wd=1.0in]|Q[l,m,wd=6.0in]| },
  rows={ rowsep=9pt },
  % set "row{1}" after "colspec" to overwrite "halign"
  row{1}={ gray!10, halign=c, font=\bfseries },
}
  \hline
  Severity & Description \\
  \hline
  a & b \\ \hline
\fileInput{in-file.tex}
\end{longtblr}
\subsection{Test Network} \label{SIL-3079}
\end{document}
```

**Reply by muzimuzhi** [reply]
> I added `\listfiles` to the preamble but couldn't find ref to the version of `tabularray` in the Console output or Error, warnings, Badboxes window.

You need to open the raw `.log` file. `\listfiles` output is not filtered out by (most) Editors.

**Reply by russurquhart1** [reply]
I'm on a Windows machine, where is the `.log` file written. I could not find it in the current directory where the LaTeX source is.

**Reply by muzimuzhi** [reply]
If your LaTeX source is named `main.tex`, then usually the log file is named `main.log` which lives in the same directory as `main.tex`.

**Reply by russurquhart1** [reply]
I just looked and there is no `.log` file for the source .tex file. Is it because it's failing at compile time?

**Reply by russurquhart1** [reply]
I looked in another log file that I had generated. Here is what it said:

`Package: tabularray 2023-03-01 v2023A Typeset tabulars and arrays with LaTeX3
`

**Reply by muzimuzhi** [reply]
The full error message you pasted was copied from the `.log` file so the log file indeed existed. Usually you can just scroll down in that tab/window to find `\listfiles` logging lines.

**Reply by muzimuzhi** [reply]
`tabularray 2023-03-01 v2023A` should be new enough, as Overleaf's TeX Live 2023 contains the identical version. By asking for `\listfiles` output I'd like to know versions of all loaded packages... .

Another perspective, what's the content of your `in-file.tex`? It should only contain `\hyperref... \\ \hline`, but never the `filecontents*` environment.

**Reply by russurquhart1** [reply]
I had put the `filecontents*` commands around it to see if that would work, I'll take it out again. Other than `tabularray` here are the only other packages that are used in that document:

```
Package: xcolor 2022/06/12 v2.14 LaTeX color extensions (UK)
Package: hyperref 2023-07-08 v7.01b Hypertext links for LaTeX
```

**Reply by russurquhart1** [reply]
HEY, it started working. After I took out the `filecontents*` stuff, It all worked!!

Thanks for your time and patience with me!! You rock!!!

**Reply by muzimuzhi** [reply]
The `filecontents(*)` environment is designed to write environment content to new files. For example in https://github.com/lvjr/tabularray/discussions/496#discussioncomment-9168895 I used `filecontents*` in the _main_ tex file to create `in-file.tex`.

See https://latexref.xyz/filecontents.html for more info.

**Reply by russurquhart1** [reply]
Very cool! Thanks again for all your help!!

**Reply by russurquhart1** [reply]
Well I'm not home yet. I tried using this library function, in my large LaTex document. The preamble is in the main file, and the array having the file to include is in another file. 
I'm getting the following error:
```
! Missing \endcsname inserted.
<to be read again> 
                   \protect 
l.32 \end
         {longtblr}

```
Here is my sample file:

```
\documentclass[landscape]{article}
\usepackage{tabularray}
\usepackage{xcolor}
\usepackage[hyperindex,hidelinks]{hyperref}
\UseTblrLibrary{functional} % Use this tabularray function to include files into a table
%\listfiles


\begin{document}
 \begin{longtblr}[
  evaluate=\fileInput, 
  caption = {Horta Errata Index}
]{ 
  rowhead = 1,
  colspec ={ |Q[c,m,wd=.75in]|Q[l,m,wd=2.7in]|Q[c,m,wd=.75in]|Q[c,m,wd=.55in]|Q[c,m,wd=.5in]|Q[c,m,wd=.55in]|Q[c,m,wd=.4in]|},
  rows={ rowsep=9pt }, 
  row{1}={ gray!10, halign=c, font=\bfseries }, 
}
  \hline
  Errata ID & %
  Title  & %
  Functional Area & %
  Category & %
%  \textbf{Severity} & %
  Work Around & %
  Stepping & %
  Ref \\ \hline



\fileInput{Index\_list.tex}
\end{longtblr}


```
And inside of `Index_list` is:

`\hyperref[SIL-3079]{SIL-3079} & Problem found & System error & A & TBD & A0 &  \ref{SIL-3079}  \\ \hline
`
I'm not seeing how this is different from the earlier sample that WAS working?

**Reply by sgolovan** [reply]
Replace `\fileInput{Index\_list.tex}` by `\fileInput{Index_list.tex}`. Also, small tip for the future: avoid using LaTeX special characters (including spaces) in file and directory names.

**Reply by russurquhart1** [reply]
Thank you! That was what did the trick. Thanks to all of you for your patience. (I wish the error that was generated had been a little more clear!) THX!!!

**Comment by russurquhart1** [comment]
I also edited my input file to look like this:

```
\begin{filecontents*}[overwrite]{in-file.tex}
\hyperref[SIL-3079]{SIL-3079} & \ref{SIL-3079}  \\ \hline
\end{filecontents*}
```
